service/history/chasm_engine.go

```diff
@@ -602,3 +609,292 @@ func (e *ChasmEngine) getExecutionLease(
+		}
+
+		valueType, ok := typeMap[fieldName]
+		if !ok {
+			return nil, fmt.Errorf("no type found for field %q", fieldName)
+		}
+
+		// Decode based on type
+		var visValue chasm.VisibilityValue
+		switch valueType {
+		case enumspb.INDEXED_VALUE_TYPE_INT:
+			var v int64
+			if err := payload.Decode(p, &v); err != nil {
+				return nil, fmt.Errorf("failed to decode int value for %q: %w", alias, err)
+			}
+			visValue = chasm.VisibilityValueInt64(v)
+		case enumspb.INDEXED_VALUE_TYPE_DOUBLE:
+			var v float64
+			if err := payload.Decode(p, &v); err != nil {
+				return nil, fmt.Errorf("failed to decode double value for %q: %w", alias, err)
+			}
+			visValue = chasm.VisibilityValueFloat64(v)
+		case enumspb.INDEXED_VALUE_TYPE_BOOL:
+			var v bool
+			if err := payload.Decode(p, &v); err != nil {
+				return nil, fmt.Errorf("failed to decode bool value for %q: %w", alias, err)
+			}
+			visValue = chasm.VisibilityValueBool(v)
+		case enumspb.INDEXED_VALUE_TYPE_DATETIME:
+			var v time.Time
+			if err := payload.Decode(p, &v); err != nil {
+				return nil, fmt.Errorf("failed to decode datetime value for %q: %w", alias, err)
+			}
+			visValue = chasm.VisibilityValueTime(v)
+		case enumspb.INDEXED_VALUE_TYPE_KEYWORD:
+			var v string
+			if err := payload.Decode(p, &v); err != nil {
+				return nil, fmt.Errorf("failed to decode keyword value for %q: %w", alias, err)
+			}
+			visValue = chasm.VisibilityValueString(v)
+		case enumspb.INDEXED_VALUE_TYPE_KEYWORD_LIST:
+			var v []string
+			if err := payload.Decode(p, &v); err != nil {
+				return nil, fmt.Errorf("failed to decode keyword list value for %q: %w", alias, err)
+			}
+			visValue = chasm.VisibilityValueStringSlice(v)
+		default:
+			return nil, fmt.Errorf("unsupported visibility value type %v for %q", valueType, alias)
+		}
+
+		result[alias] = visValue
+	}
+
+	return result, nil
+}
+
+// splitSearchAttributes separates search attributes into CHASM and custom search attributes.
+// At this point, search attributes are already aliased (both CHASM and custom).
+// We just need to determine which ones are CHASM search attributes.
+func splitSearchAttributes(
+	searchAttributes *commonpb.SearchAttributes,
+	mapper *chasm.VisibilitySearchAttributesMapper,
+) (chasmSearchAttributes, customSearchAttributes map[string]*commonpb.Payload) {
+	if searchAttributes == nil || len(searchAttributes.IndexedFields) == 0 {
+		return nil, nil
+	}
+
+	chasmSAs := make(map[string]*commonpb.Payload)
+	customSAs := make(map[string]*commonpb.Payload)
+
+	for saName, saPayload := range searchAttributes.IndexedFields {
+		// Check if this is a CHASM search attribute by trying to get its field name
+		// If the mapper knows about this alias, it's a CHASM search attribute
+		if mapper != nil {
+			if _, err := mapper.Field(saName); err == nil {
+				chasmSAs[saName] = saPayload
+				continue
+			}
+		}
+
+		// Otherwise, it's a custom search attribute
+		customSAs[saName] = saPayload
+	}
```

Comment:
```suggestion
	if mapper == nil {
		...
	}
	for saName, saPayload := range searchAttributes.IndexedFields {
		// Check if this is a CHASM search attribute by trying to get its field name
		// If the mapper knows about this alias, it's a CHASM search attribute
		if _, err := mapper.Field(saName); err == nil {
			chasmSAs[saName] = saPayload
			continue
		}

		// Otherwise, it's a custom search attribute
		customSAs[saName] = saPayload
	}
```